test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts

```diff
@@ -94,6 +90,88 @@ describe('ppr-navigations', () => {
     const loading = await browser.elementById('loading-tsx')
     expect(await loading.innerHTML()).toEqual('Loading [inner loading.tsx]...')
   })
+
+  // This is a regression test where a condition that checked whether a cached
+  // route entry was stale accidentally caused the router to treat prefetched
+  // data as if it were complete data that didn't contain dynamic holes. This
+  // prevented the dynamic data from ever streaming in.
+  test(
+    'works if a prefetched route entry has become stale (too much ' +
+      'time has elapsed since it was prefetched)',
+    async () => {
+      const TestLog = createTestLog()
+      let autoresolveRequests = true
+      let pendingRequests = new Map()
+      server = createTestDataServer(async (key, res) => {
+        TestLog.log('REQUEST: ' + key)
+        if (autoresolveRequests) {
+          res.resolve()
+          return
+        }
+        if (pendingRequests.has(key)) {
+          throw new Error('Request already pending for ' + key)
+        }
+        pendingRequests.set(key, res)
+      })
+      const port = await findPort()
+      server.listen(port)
+      next = await createNext({
+        files: __dirname + '/stale-prefetch-entry',
+        env: { TEST_DATA_SERVICE_URL: `http://localhost:${port}` },
+      })
+      TestLog.assert(['REQUEST: Some data [static]'])
+      autoresolveRequests = false
+
+      const browser = await next.browser('/')
+
+      // Override Date.now so we can simulate time passing to expire a
+      // prefetch entry.
+      const page = browser.getPlayrightPage()
+      await page.addInitScript(`
+      __FAKE_NOW__ = 0
+      Date = class extends Date {
+        constructor(...args) {
+          if (args.length === 0) {
+            super(__FAKE_NOW__)
+          } else {
+            super(...args)
+          }
+        }
+        static now() {
+          return __FAKE_NOW__
+        }
+      }
+    `)
```

Comment:
This was the least bad way I could find to simulate time elapsing. Turns out this is a known gap in Playwright: https://github.com/microsoft/playwright/issues/6347

Reply:
did you the logic I added for the client-cache tests as well?

Reply:
kinda doing the same things

Reply:
Oh no I didn't but I just looked it up, that's neat that we independently used the same pattern.